File: app/src/main/java/com/khmelenko/lab/varis/activity/SearchResultsActivity.java
package com.khmelenko.lab.varis.activity;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.khmelenko.lab.varis.R;
import com.khmelenko.lab.varis.TravisApp;
import com.khmelenko.lab.varis.fragment.ReposFragment;
import com.khmelenko.lab.varis.mvp.MvpActivity;
import com.khmelenko.lab.varis.network.response.Repo;
import com.khmelenko.lab.varis.presenter.SearchResultsPresenter;
import com.khmelenko.lab.varis.view.SearchResultsView;

import java.util.List;

import javax.inject.Inject;

import butterknife.ButterKnife;

/**
 * Provides activity with search results
 *
 * @author Dmytro Khmelenko
 */
public final class SearchResultsActivity extends MvpActivity<SearchResultsPresenter> implements
        SearchResultsView,
        ReposFragment.ReposFragmentListener {

    @Inject
    SearchResultsPresenter mPresenter;

    private ReposFragment mFragment;

    private String mSearchQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_results);
        TravisApp.instance().activityInjector().inject(this);
        ButterKnife.bind(this);

        mFragment = (ReposFragment) getFragmentManager().findFragmentById(R.id.search_fragment_repos);

        initToolbar();

        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        setIntent(intent);
        handleIntent(intent);
    }

    @Override
    protected SearchResultsPresenter getPresenter() {
        return mPresenter;
    }

    @Override
    protected void attachPresenter() {
        getPresenter().attach(this);
    }

    /**
     * Handles received intent
     *
     * @param intent Intent
     */
    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            String query = intent.getStringExtra(SearchManager.QUERY);
            mSearchQuery = query;
            showProgress();
            getPresenter().startRepoSearch(query);
        }
    }

    /**
     * Initializes toolbar
     */
    private void initToolbar() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
            toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onBackPressed();
                }
            });
        }
    }

    @Override
    public void onRepositorySelected(Repo repo) {
        Intent intent = new Intent(SearchResultsActivity.this, RepoDetailsActivity.class);
        intent.putExtra(RepoDetailsActivity.REPO_SLUG_KEY, repo.getSlug());
        startActivity(intent);
    }

    @Override
    public void onRefreshData() {
        getPresenter().startRepoSearch(mSearchQuery);
    }

    @Override
    public void showProgress() {
        mFragment.setLoadingProgress(true);
    }

    @Override
    public void hideProgress() {
        mFragment.setLoadingProgress(false);
    }

    @Override
    public void setSearchResults(List<Repo> repos) {
        mFragment.setRepos(repos);
    }

    @Override
    public void showLoadingError(String message) {
        mFragment.handleLoadingFailed(message);
    }
}

File: app-v3/src/main/java/com/khmelenko/lab/varis/repositories/search/SearchResultsActivity.kt
package com.khmelenko.lab.varis.repositories.search

import android.app.SearchManager
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.Toolbar
import com.khmelenko.lab.varis.R
import com.khmelenko.lab.varis.activity.BaseActivity
import com.khmelenko.lab.varis.network.response.Repo
import com.khmelenko.lab.varis.repodetails.REPO_SLUG_KEY
import com.khmelenko.lab.varis.repodetails.RepoDetailsActivity
import com.khmelenko.lab.varis.repodetails.RepoDetailsViewModelFactory
import com.khmelenko.lab.varis.repositories.ReposFragment
import dagger.android.AndroidInjection
import javax.inject.Inject

/**
 * Provides activity with search results
 *
 * @author Dmytro Khmelenko
 */
class SearchResultsActivity : BaseActivity(), ReposFragment.ReposFragmentListener {

    @Inject
    lateinit var viewModelFactory: RepoDetailsViewModelFactory

    private lateinit var viewModel: SearchResultsViewModel

    private lateinit var fragment: ReposFragment

    private var searchQuery: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_results)

        viewModel = ViewModelProviders.of(this, viewModelFactory).get(SearchResultsViewModel::class.java)
        viewModel.state().observe(this, Observer { state ->
            hideProgress()
            when (state) {
                is SearchState.Loading -> showProgress()
                is SearchState.ReposList -> setSearchResults(state.repos)
                is SearchState.Error -> showLoadingError(state.message)
            }
        })

        fragment = fragmentManager.findFragmentById(R.id.search_fragment_repos) as ReposFragment

        initToolbar()

        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        setIntent(intent)
        handleIntent(intent)
    }

    /**
     * Handles received intent
     *
     * @param intent Intent
     */
    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            val query = intent.getStringExtra(SearchManager.QUERY)
            searchQuery = query
            showProgress()
            viewModel.startRepoSearch(query)
        }
    }

    /**
     * Initializes toolbar
     */
    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowHomeEnabled(true)
            toolbar.setNavigationOnClickListener { v -> onBackPressed() }
        }
    }

    override fun onRepositorySelected(repo: Repo) {
        val intent = Intent(this@SearchResultsActivity, RepoDetailsActivity::class.java)
        intent.putExtra(REPO_SLUG_KEY, repo.slug)
        startActivity(intent)
    }

    override fun onRefreshData() {
        viewModel.startRepoSearch(searchQuery!!)
    }

    private fun showProgress() {
        fragment.setLoadingProgress(true)
    }

    private fun hideProgress() {
        fragment.setLoadingProgress(false)
    }

    private fun setSearchResults(repos: List<Repo>) {
        fragment.setRepos(repos)
    }

    private fun showLoadingError(message: String?) {
        fragment.handleLoadingFailed(message)
    }
}
